File: projects/js/tests/TestDoor.js
import {StaterStateMachine, StaterStateMachineBuilder} from "../src/index.js";

const States = Object.freeze({
    CLOSE: "CLOSE",
    AJAR: "AJAR",
    OPEN: "OPEN"
});

const getState = (str) => {
    if (Object.values(States).includes(str)) {
        return str;
    }
    throw new Error("Invalid state");
};

let buildDoorDSMContext = () => {
    return {degreeOfOpening: 100}
}


class TypesDoorStateMachine extends StaterStateMachine {
    constructor(
        transitions,
        startState,
        states,
        context,
        transitionMiddlewares = new Map(),
        transitionAllMiddlewares = [],
        transitionCallbacks = new Map(),
        transitionAllCallbacks = [],
        stateCallbacks = new Map(),
        stateAllCallbacks = [],
        contextJsonAdapter = undefined
    ) {
        super(
            transitions,
            startState,
            states,
            context,
            transitionMiddlewares,
            transitionAllMiddlewares,
            transitionCallbacks,
            transitionAllCallbacks,
            stateCallbacks,
            stateAllCallbacks,
            contextJsonAdapter
        );
    }

    ajarPlus() {
        this.transition("ajarPlus");
    }

    ajarMinus() {
        this.transition("ajarMinus");
    }

    preOpen() {
        this.transition("preOpen");
    }

    preClose() {
        this.transition("preClose");
    }

    open() {
        this.transition("open");
    }

    close() {
        this.transition("close");
    }
}


const typedDoorFactory = (
    transitionsA,
    startStateA,
    statesA,
    contextA,
    transitionMiddlewaresA,
    transitionAllMiddlewaresA,
    transitionCallbacksA,
    transitionAllCallbacksA,
    stateCallbacksA,
    stateAllCallbacksA,
    contextJsonAdapterA
) => {
    return new TypesDoorStateMachine(
        transitionsA,
        startStateA,
        statesA,
        contextA,
        transitionMiddlewaresA,
        transitionAllMiddlewaresA,
        transitionCallbacksA,
        transitionAllCallbacksA,
        stateCallbacksA,
        stateAllCallbacksA,
        contextJsonAdapterA
    );
};


function testDoor(door) {
    console.assert(door.getState() === States.OPEN);
    console.assert(door.getContext().degreeOfOpening === 100);
    door.transition("preClose");
    console.assert(door.getState() === States.AJAR);
    console.assert(door.getContext().degreeOfOpening === 99);
    while (door.getContext().degreeOfOpening > 1) {
        door.transition("ajarMinus");
        console.assert(door.getState() === States.AJAR);
    }
    console.assert(door.getContext().degreeOfOpening === 1);
    door.transition("close");
    console.assert(door.getContext().degreeOfOpening === 0);
    console.assert(door.getState() === States.CLOSE);
    door.transition("preOpen");
    console.assert(door.getContext().degreeOfOpening === 1);
    console.assert(door.getState() === States.AJAR);
    door.transition("ajarPlus");
    console.assert(door.getState() === States.AJAR);
    console.assert(door.getContext().degreeOfOpening === 2);
    while (door.getContext().degreeOfOpening < 99) {
        door.transition("ajarPlus");
        console.assert(door.getState() === States.AJAR);
    }
    door.transition("open");
    console.assert(door.getState() === States.OPEN);
    console.assert(door.getContext().degreeOfOpening === 100);
}

function typedTestDoor(door) {
    console.assert(door.getState() === States.OPEN);
    console.assert(door.getContext().degreeOfOpening === 100);
    door.preClose();
    console.assert(door.getState() === States.AJAR);
    console.assert(door.getContext().degreeOfOpening === 99);
    while (door.getContext().degreeOfOpening > 1) {
        door.ajarMinus();
        console.assert(door.getState() === States.AJAR);
    }
    console.assert(door.getContext().degreeOfOpening === 1);
    door.close();
    console.assert(door.getContext().degreeOfOpening === 0);
    console.assert(door.getState() === States.CLOSE);
    door.preOpen();
    console.assert(door.getContext().degreeOfOpening === 1);
    console.assert(door.getState() === States.AJAR);
    door.ajarPlus();
    console.assert(door.getState() === States.AJAR);
    console.assert(door.getContext().degreeOfOpening === 2);
    while (door.getContext().degreeOfOpening < 99) {
        door.ajarPlus();
        console.assert(door.getState() === States.AJAR);
    }
    door.open();
    console.assert(door.getState() === States.OPEN);
    console.assert(door.getContext().degreeOfOpening === 100);
}

function testSimpleBuild() {
    const doorFSM = new TypesDoorStateMachine([
            {
                name: "preOpen",
                start: States.CLOSE,
                end: States.AJAR,
                event: (context) => {
                    context.degreeOfOpening = 1;
                }
            },
            {
                name: "preClose",
                start: States.OPEN,
                end: States.AJAR,
                event: (context) => {
                    context.degreeOfOpening = 99;
                }
            },
            {
                name: "open",
                start: States.AJAR,
                end: States.OPEN,
                condition: (context) => context.degreeOfOpening >= 99,
                event: (context) => {
                    context.degreeOfOpening = 100;
                }
            },
            {
                name: "close",
                start: States.AJAR,
                end: States.CLOSE,
                condition: (context) => context.degreeOfOpening <= 1,
                event: (context) => {
                    context.degreeOfOpening = 0;
                }
            },
            {
                name: "ajarPlus",
                start: States.AJAR,
                end: States.AJAR,
                condition: (context) => context.degreeOfOpening >= 1 && context.degreeOfOpening <= 98,
                event: (context) => {
                    context.degreeOfOpening++;
                }
            },
            {
                name: "ajarMinus",
                start: States.AJAR,
                end: States.AJAR,
                condition: (context) => context.degreeOfOpening >= 2 && context.degreeOfOpening <= 99,
                event: (context) => {
                    context.degreeOfOpening--;
                }
            }
        ],
        States.OPEN,
        [States.OPEN, States.CLOSE, States.AJAR],
        buildDoorDSMContext(),
    );

    testDoor(doorFSM);
    typedTestDoor(doorFSM);
}


function testBuilder() {
    const doorFSM = new StaterStateMachineBuilder()
        .addTransition("preOpen", States.CLOSE, States.AJAR, undefined, (context) => {
            context.degreeOfOpening = 1
        })
        .addTransition("preClose", States.OPEN, States.AJAR, undefined, (context) => {
            context.degreeOfOpening = 99
        })
        .addTransition("open", States.AJAR, States.OPEN, (context) => context.degreeOfOpening >= 99, (context) => {
            context.degreeOfOpening = 100
        })
        .addTransition("close", States.AJAR, States.CLOSE, (context) => context.degreeOfOpening <= 1, (context) => {
            context.degreeOfOpening = 0
        })
        .addTransition("ajarPlus", States.AJAR, States.AJAR, (context) => context.degreeOfOpening >= 1 && context.degreeOfOpening <= 98, (context) => {
            context.degreeOfOpening++
        })
        .addTransition("ajarMinus", States.AJAR, States.AJAR, (context) => context.degreeOfOpening >= 2 && context.degreeOfOpening <= 99, (context) => {
            context.degreeOfOpening--
        })
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .build();

    testDoor(doorFSM);
}

function structureBuild() {
    return new StaterStateMachineBuilder()
        .addTransition("preOpen", States.CLOSE, States.AJAR)
        .addTransition("preClose", States.OPEN, States.AJAR)
        .addTransition("open", States.AJAR, States.OPEN)
        .addTransition("close", States.AJAR, States.CLOSE)
        .addTransition("ajarPlus", States.AJAR, States.AJAR)
        .addTransition("ajarMinus", States.AJAR, States.AJAR)
}

function eventsBuild(builder) {
    return builder
        .setTransitionEvent("preOpen", (context) => {
            context.degreeOfOpening = 1
        })
        .setTransitionEvent("preClose", (context) => {
            context.degreeOfOpening = 99
        })
        .setTransitionCondition("open", (context) => context.degreeOfOpening >= 99)
        .setTransitionEvent("open", (context) => {
            context.degreeOfOpening = 100
        })
        .setTransitionCondition("close", (context) => context.degreeOfOpening <= 1)
        .setTransitionEvent("close", (context) => {
            context.degreeOfOpening = 0
        })
        .setTransitionCondition("ajarPlus", (context) => context.degreeOfOpening >= 1 && context.degreeOfOpening <= 98)
        .setTransitionEvent("ajarPlus", (context) => {
            context.degreeOfOpening++
        })
        .setTransitionCondition("ajarMinus", (context) => context.degreeOfOpening >= 2 && context.degreeOfOpening <= 99)
        .setTransitionEvent("ajarMinus", (context) => {
            context.degreeOfOpening--
        });
}


function testBuilder2() {
    const doorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .build()

    testDoor(doorFSM)
}

function testAutoTransition() {
    const doorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .build()

    doorFSM.autoTransition()
    console.assert(doorFSM.getState() === States.AJAR)
}


function testBuilderFactory() {
    const doorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .setFactory(typedDoorFactory)
        .build();

    testDoor(doorFSM);
    console.assert(doorFSM instanceof TypesDoorStateMachine)
    if (doorFSM instanceof TypesDoorStateMachine) {
        typedTestDoor(doorFSM);
    }
}

function testJsonSchema() {
    const validDoorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .build();

    testDoor(validDoorFSM);

    const jsonSchema = validDoorFSM.toJsonSchema();

    const doorFSM = eventsBuild(
        structureBuild()
            .fromJsonSchema(jsonSchema, getState))
        .setContext(buildDoorDSMContext())
        .setFactory(typedDoorFactory)
        .build();

    console.assert(JSON.stringify(jsonSchema) === JSON.stringify(doorFSM.toJsonSchema()));

    testDoor(doorFSM);

    if (doorFSM instanceof TypesDoorStateMachine) {
        typedTestDoor(doorFSM);
    }
}

function testStringGeneric() {
    const validDoorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .build();

    testDoor(validDoorFSM);

    const jsonSchema = validDoorFSM.toJsonSchema();

    const stringDoorFSM = new StaterStateMachineBuilder()
        .fromJsonSchema(jsonSchema, (stateStr) => stateStr)
        .setContext(buildDoorDSMContext())
        .build();

    console.assert(stringDoorFSM.getState() === "OPEN");

}

function testJsonDump() {
    class JsonConverter {
        toJson(context) {
            return context.degreeOfOpening.toString();
        }

        fromJson(json) {
            return {degreeOfOpening: parseInt(json)};
        }
    }

    const validDoorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .setContextJsonAdapter(new JsonConverter())
        .build();

    console.assert(validDoorFSM.getState() === States.OPEN);
    console.assert(validDoorFSM.getContext().degreeOfOpening === 100);

    const dump = validDoorFSM.toJson();

    validDoorFSM.transition("preClose");
    console.assert(validDoorFSM.getState() === States.AJAR);
    console.assert(validDoorFSM.getContext().degreeOfOpening === 99);

    validDoorFSM.fromJson(dump, getState);
    console.assert(validDoorFSM.getState() === States.OPEN);
    console.assert(validDoorFSM.getContext().degreeOfOpening === 100);

}

function testMiddlewareAndCallbacks() {
    let transitionMiddleware = 0;
    let transitionAllMiddleware = 0;
    let subscribeOnTransition = 0;
    let subscribeOnAllTransition = 0;
    let subscribeOnState = 0;
    let subscribeOnAllState = 0;

    const validDoorFSM = eventsBuild(structureBuild())
        .setContext(buildDoorDSMContext())
        .setStartState(States.OPEN)
        .transitionMiddleware("open", (context, next) => {
            transitionMiddleware++;
            next(context);
        })
        .transitionAllMiddleware((name, context, next) => {
            transitionAllMiddleware++;
            next(name, context);
        })
        .subscribeOnTransition("open", () => {
            subscribeOnTransition++;
        })
        .subscribeOnAllTransition((name, context) => {
            subscribeOnAllTransition++;
        })
        .subscribeOnState(States.AJAR, () => {
            subscribeOnState++;
        })
        .subscribeOnAllState((name, context) => {
            subscribeOnAllState++;
        })
        .build();

    testDoor(validDoorFSM);

    console.assert(transitionMiddleware === 1);
    console.assert(transitionAllMiddleware === 200);
    console.assert(subscribeOnTransition === 1);
    console.assert(subscribeOnAllTransition === 200);
    console.assert(subscribeOnState === 198);
    console.assert(subscribeOnAllState === 200);
}


testSimpleBuild()
testBuilder()
testBuilder2()
testAutoTransition()
testBuilderFactory()
testJsonSchema()
testStringGeneric()
testJsonDump()
testMiddlewareAndCallbacks()

File: projects/c++/tests/test_door.cpp
//
// Created by Владимир on 02.04.2025.
//

#include <stater_state_machine.h>
#include <gtest/gtest.h>

#include <utility>

enum class states {
    CLOSE,
    AJAR,
    OPEN
};

class door_fsm_context : context {
    int degree_of_opening;

public:
    door_fsm_context() : degree_of_opening(100) {
    }

    explicit door_fsm_context(int degree_of_opening) : degree_of_opening(degree_of_opening) {
    }

    [[nodiscard]] int get_degree_of_opening() const {
        return degree_of_opening;
    }

    void set_degree_of_opening(int degree_of_opening) {
        this->degree_of_opening = degree_of_opening;
    }
};

door_fsm_context door_fsm_context_ = door_fsm_context();

class types_door_state_machine : public stater_state_machine<states, door_fsm_context> {
public:
    types_door_state_machine() = default;

    // : stater_state_machine(
        //     {
        //         // Transition<::states, door_fsm_context>("preOpen", states::CLOSE, states::AJAR,
        //         //                                        [](door_fsm_context *ctx) { return true; },
        //         //                                        [](door_fsm_context *ctx) { ctx->set_degree_of_opening(1); }),
        //         // Transition<::states, door_fsm_context>("preClose", states::OPEN, states::AJAR,
        //         //                                        [](door_fsm_context *ctx) { return true; },
        //         //                                        [](door_fsm_context *ctx) { ctx->set_degree_of_opening(99); }),
        //         // Transition<::states, door_fsm_context>("open", states::AJAR, states::OPEN,
        //         //                                        [](door_fsm_context *ctx) {
        //         //                                            return ctx->get_degree_of_opening() >= 99;
        //         //                                        },
        //         //                                        [](door_fsm_context *ctx) { ctx->set_degree_of_opening(100); }),
        //         // Transition<::states, door_fsm_context>("close", states::AJAR, states::CLOSE,
        //         //                                        [](door_fsm_context *ctx) {
        //         //                                            return ctx->get_degree_of_opening() <= 1;
        //         //                                        },
        //         //                                        [](door_fsm_context *ctx) { ctx->set_degree_of_opening(0); }),
        //         // Transition<::states, door_fsm_context>("ajarPlus", states::AJAR, states::AJAR,
        //         //                                        [](door_fsm_context *ctx) {
        //         //                                            return ctx->get_degree_of_opening() >= 1 && ctx->
        //         //                                                   get_degree_of_opening() <= 98;
        //         //                                        },
        //         //                                        [](door_fsm_context *ctx) {
        //         //                                            ctx->set_degree_of_opening(ctx->get_degree_of_opening() + 1);
        //         //                                        }),
        //         // Transition<::states, door_fsm_context>("ajarMinus", states::AJAR, states::AJAR,
        //         //                                        [](door_fsm_context *ctx) {
        //         //                                            return ctx->get_degree_of_opening() >= 2 && ctx->
        //         //                                                   get_degree_of_opening() <= 99;
        //         //                                        },
        //         //                                        [](door_fsm_context *ctx) {
        //         //                                            ctx->set_degree_of_opening(ctx->get_degree_of_opening() - 1);
        //         //                                        })
        //     },
        //     &door_fsm_context_,
        //     ::states::OPEN,
        //     {::states::OPEN, ::states::CLOSE, ::states::AJAR},
        //     {},
        //     {},
        //     {},
        //     {},
        //     {},
        //     {}
        // ) {

    void ajar_plus() {
        transition("ajarPlus");
    }

    void ajar_minus() {
        transition("ajarMinus");
    }

    void pre_open() {
        transition("preOpen");
    }

    void pre_close() {
        transition("preClose");
    }

    void open() {
        transition("open");
    }

    void close() {
        transition("close");
    }
};

void test_door_schema(stater_state_machine<states, door_fsm_context> door) {
    // ASSERT_EQ(states::OPEN, door.get_state());
    // ASSERT_EQ(100, door.get_context().get_degree_of_opening());
    // door.transition("preClose");
    // ASSERT_EQ(states::AJAR, door.get_state());
    // ASSERT_EQ(99, door.get_context().get_degree_of_opening());
    // while (door.get_context().get_degree_of_opening() > 1) {
    //     door.transition("ajarMinus");
    //     ASSERT_EQ(states::AJAR, door.get_state());
    // }
    //
    // ASSERT_EQ(1, door.get_context().get_degree_of_opening());
    // door.transition("close");
    // ASSERT_EQ(0, door.get_context().get_degree_of_opening());
    // ASSERT_EQ(states::CLOSE, door.get_state());
    // door.transition("preOpen");
    // ASSERT_EQ(1, door.get_context().get_degree_of_opening());
    // ASSERT_EQ(states::AJAR, door.get_state());
    // door.transition("ajarPlus");
    // ASSERT_EQ(states::AJAR, door.get_state());
    // ASSERT_EQ(2, door.get_context().get_degree_of_opening());
    // while (door.get_context().get_degree_of_opening() < 99) {
    //     door.transition("ajarPlus");
    //     ASSERT_EQ(states::AJAR, door.get_state());
    // }
    //
    // door.transition("open");
    // ASSERT_EQ(states::OPEN, door.get_state());
    // ASSERT_EQ(100, door.get_context().get_degree_of_opening());
}

void typed_test_schema(types_door_state_machine door) {
    // ASSERT_EQ(states::OPEN, door.get_state());
    // ASSERT_EQ(100, door.get_context().get_degree_of_opening());
    // door.pre_close();
    // ASSERT_EQ(states::AJAR, door.get_state());
    // ASSERT_EQ(99, door.get_context().get_degree_of_opening());
    // while (door.get_context().get_degree_of_opening() > 1) {
    //     door.ajar_minus();
    //     ASSERT_EQ(states::AJAR, door.get_state());
    // }
    //
    // ASSERT_EQ(1, door.get_context().get_degree_of_opening());
    // door.close();
    // ASSERT_EQ(0, door.get_context().get_degree_of_opening());
    // ASSERT_EQ(states::CLOSE, door.get_state());
    // door.pre_open();
    // ASSERT_EQ(1, door.get_context().get_degree_of_opening());
    // ASSERT_EQ(states::AJAR, door.get_state());
    // door.ajar_plus();
    // ASSERT_EQ(states::AJAR, door.get_state());
    // ASSERT_EQ(2, door.get_context().get_degree_of_opening());
    // while (door.get_context().get_degree_of_opening() < 99) {
    //     door.ajar_plus();
    //     ASSERT_EQ(states::AJAR, door.get_state());
    // }
    //
    // door.open();
    // ASSERT_EQ(states::OPEN, door.get_state());
    // ASSERT_EQ(100, door.get_context().get_degree_of_opening());
}


TEST(state_machine_test, test_simple_build) {
    auto door_fsm = types_door_state_machine();
}

TEST(state_machine_test, test_builder) {
    // door_fsm_context context;
    //
    // auto door_fsm = stater_state_machine_builder<states, door_fsm_context>()
    //         .add_transition("preOpen", states::CLOSE, states::AJAR,
    //                         [](door_fsm_context *ctx) { return true; },
    //                         [](door_fsm_context *ctx) { ctx->set_degree_of_opening(1); })
    //         .add_transition("preClose", states::OPEN, states::AJAR,
    //                         [](door_fsm_context *ctx) { return true; },
    //                         [](door_fsm_context *ctx) { ctx->set_degree_of_opening(99); })
    //         .add_transition("open", states::AJAR, states::OPEN,
    //                         [](door_fsm_context *ctx) { return ctx->get_degree_of_opening() >= 99; },
    //                         [](door_fsm_context *ctx) { ctx->set_degree_of_opening(100); })
    //         .add_transition("close", states::AJAR, states::CLOSE,
    //                         [](door_fsm_context *ctx) { return ctx->get_degree_of_opening() <= 1; },
    //                         [](door_fsm_context *ctx) { ctx->set_degree_of_opening(0); })
    //         .add_transition("ajarPlus", states::AJAR, states::AJAR,
    //                         [](door_fsm_context *ctx) {
    //                             return 1 <= ctx->get_degree_of_opening() && ctx->get_degree_of_opening() <= 98;
    //                         },
    //                         [](door_fsm_context *ctx) { ctx->set_degree_of_opening(ctx->get_degree_of_opening() + 1); })
    //         .add_transition("ajarMinus", states::AJAR, states::AJAR,
    //                         [](door_fsm_context *ctx) {
    //                             return 2 <= ctx->get_degree_of_opening() && ctx->get_degree_of_opening() <= 99;
    //                         },
    //                         [](door_fsm_context *ctx) { ctx->set_degree_of_opening(ctx->get_degree_of_opening() - 1); })
    //         .set_context(&context)
    //         .set_start_state(states::OPEN)
    //         .build();
    //
    // test_door_schema(door_fsm);
}

TEST(state_machine_test, TestBuilder2) {
}

TEST(state_machine_test, TestAutoTransition) {
}

TEST(state_machine_test, TestBuilderFactory) {
}

TEST(state_machine_test, TestJsonSchema) {
}

TEST(state_machine_test, TestStringGeneric) {
}

TEST(state_machine_test, TestJsonDump) {
}

TEST(state_machine_test, TestMiddlewareAndCallbacks) {
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
